File: packages/py/tracing/examples/langgraph.py
from typing_extensions import TypedDict
from langgraph.graph import END, START, StateGraph
from uselemma_tracing import langgraph


class GraphState(TypedDict):
    input: str
    output: str


def answer(state: GraphState):
    return {"output": f"You said: {state['input']}"}


graph = (
    StateGraph(GraphState)
    .add_node("answer", answer)
    .add_edge(START, "answer")
    .add_edge("answer", END)
    .compile()
)

# langgraph() is a LangChain callback adapter with LangGraph defaults.
lemma_handler = langgraph(
    agent_name="support-graph",
    thread_id_key="thread_id",
)


def call_langgraph(user_message: str, thread_id: str | None = None):
    config: dict = {"callbacks": [lemma_handler]}
    if thread_id:
        config["metadata"] = {"thread_id": thread_id}
    result = graph.invoke({"input": user_message}, config)
    return result["output"]


def shutdown_langgraph():
    lemma_handler.shutdown()

File: packages/ts/tracing/examples/langgraph.ts
import { StateGraph, START, END } from "@langchain/langgraph";
import { langGraph } from "@uselemma/tracing";

type GraphState = {
  input: string;
  output?: string;
};

const graph = new StateGraph<GraphState>()
  .addNode("answer", async (state) => ({
    output: `You said: ${state.input}`,
  }))
  .addEdge(START, "answer")
  .addEdge("answer", END)
  .compile();

/** langGraph() is a LangChain callback adapter with LangGraph defaults. */
const lemmaHandler = langGraph({
  agentName: "support-graph",
  threadIdKey: "thread_id",
});

export async function callLangGraph(
  userMessage: string,
  threadId?: string,
) {
  const result = await graph.invoke(
    { input: userMessage },
    {
      callbacks: [lemmaHandler],
      metadata: threadId ? { thread_id: threadId } : undefined,
    },
  );

  return result.output;
}

export async function shutdownLangGraph() {
  await lemmaHandler.shutdown();
}
